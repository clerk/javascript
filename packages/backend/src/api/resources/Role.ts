import type { RoleJSON } from './JSON';
import { Permission } from './Permission';

export class Role {
  constructor(
    readonly id: string,
    readonly name: string,
    readonly key: string,
    readonly description: string,
    readonly permissions: Permission[] = [],
    readonly createdAt: number,
    readonly updatedAt: number,
  ) {}

  static fromJSON(data: RoleJSON): Role {
    return new Role(
      data.id,
      data.name,
      data.key,
      data.description,
      (data.permissions || []).map(x => Permission.fromJSON(x)),
      data.created_at,
      data.updated_at,
    );
  }
}
