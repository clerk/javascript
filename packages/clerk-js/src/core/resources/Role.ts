import type { RoleJSON, RoleResource } from '@clerk/types';

import { unixEpochToDate } from '../../utils/date';
import { BaseResource } from './internal';
import { Permission } from './Permission';

/**
 * @experimental
 */
export class Role extends BaseResource implements RoleResource {
  id!: string;
  key!: string;
  name!: string;
  description!: string;
  permissions: Permission[] = [];
  createdAt!: Date;
  updatedAt!: Date;

  constructor(data: RoleJSON) {
    super();
    this.fromJSON(data);
  }

  protected fromJSON(data: RoleJSON | null): this {
    if (!data) {
      return this;
    }

    this.id = data.id;
    this.key = data.key;
    this.name = data.name;
    this.description = data.description;
    this.permissions = data.permissions.map(perm => new Permission(perm));
    this.createdAt = unixEpochToDate(data.created_at);
    this.updatedAt = unixEpochToDate(data.updated_at);
    return this;
  }
}
