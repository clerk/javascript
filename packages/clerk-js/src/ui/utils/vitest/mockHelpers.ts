// import { jest } from '@jest/globals';
import type { LoadedClerk } from '@clerk/types';
import type { ActiveSessionResource } from '@clerk/types';
import { type Mocked, vi } from 'vitest';

import type { RouteContextValue } from '../../router';

type FunctionLike = (...args: any) => any;

type DeepVitestMocked<T> = T extends FunctionLike
  ? Mocked<T>
  : T extends object
    ? {
        [k in keyof T]: DeepVitestMocked<T[k]>;
      }
    : T;

// Removing jest.Mock type for now, relying on inference
type MockMap<T = any> = {
  [K in { [K in keyof T]: T[K] extends (...args: any[]) => any ? K : never }[keyof T]]?: ReturnType<typeof vi.fn>;
};

const mockProp = <T>(obj: T, k: keyof T, mocks?: MockMap<T>) => {
  if (typeof obj[k] === 'function') {
    const mockFn = mocks?.[k] ?? vi.fn();
    (obj[k] as unknown as ReturnType<typeof vi.fn>) = mockFn;
  }
};

const mockMethodsOf = <T extends Record<string, any> | null = any>(
  obj: T,
  options?: {
    exclude: (keyof T)[];
    mocks: MockMap<T>;
  },
) => {
  if (!obj) {
    return;
  }
  Object.keys(obj)
    .filter(key => !options?.exclude.includes(key as keyof T))
    // Pass the specific MockMap for the object T being mocked
    .forEach(k => mockProp(obj, k, options?.mocks));
};

export const mockClerkMethods = (clerk: LoadedClerk): DeepVitestMocked<LoadedClerk> => {
  // Cast clerk to any to allow mocking properties
  const clerkAny = clerk as any;

  mockMethodsOf(clerkAny);
  if (clerkAny.client) {
    mockMethodsOf(clerkAny.client.signIn);
    mockMethodsOf(clerkAny.client.signUp);
    clerkAny.client.sessions?.forEach((session: ActiveSessionResource) => {
      const sessionAny = session as any;
      mockMethodsOf(sessionAny, {
        exclude: ['checkAuthorization'],
        mocks: {
          // Ensure touch mock matches expected signature if available, otherwise basic mock
          touch: vi.fn(() => Promise.resolve(session)),
        },
      });
      if (sessionAny.user) {
        mockMethodsOf(sessionAny.user);
        sessionAny.user.emailAddresses?.forEach((m: any) => mockMethodsOf(m));
        sessionAny.user.phoneNumbers?.forEach((m: any) => mockMethodsOf(m));
        sessionAny.user.externalAccounts?.forEach((m: any) => mockMethodsOf(m));
        sessionAny.user.organizationMemberships?.forEach((m: any) => {
          mockMethodsOf(m);
          if (m.organization) {
            mockMethodsOf(m.organization);
          }
        });
        sessionAny.user.passkeys?.forEach((m: any) => mockMethodsOf(m));
      }
    });
  }
  mockProp(clerkAny, 'navigate');
  mockProp(clerkAny, 'setActive');
  mockProp(clerkAny, 'redirectWithAuth');
  mockProp(clerkAny, '__internal_navigateWithError');
  return clerkAny as DeepVitestMocked<LoadedClerk>;
};

export const mockRouteContextValue = ({ queryString = '' }: Partial<DeepVitestMocked<RouteContextValue>>) => {
  return {
    basePath: '',
    startPath: '',
    flowStartPath: '',
    fullPath: '',
    indexPath: '',
    currentPath: '',
    queryString,
    queryParams: {},
    getMatchData: vi.fn(),
    matches: vi.fn(),
    baseNavigate: vi.fn(),
    navigate: vi.fn(() => Promise.resolve(true)),
    resolve: vi.fn((to: string) => new URL(to, 'https://clerk.com')),
    refresh: vi.fn(),
    params: {},
  } as RouteContextValue; // Keep original type assertion, DeepVitestMocked applied to input only
};
