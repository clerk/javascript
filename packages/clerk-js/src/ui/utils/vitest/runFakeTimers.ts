import { vi } from 'vitest';
// import { jest } from '@jest/globals';
import { act } from '@testing-library/react';

type WithAct = <T>(fn: T) => T;
const withAct = ((fn: any) =>
  (...args: any) => {
    act(() => {
      fn(...args);
    });
  }) as WithAct;

const advanceTimersByTime = withAct(vi.advanceTimersByTime.bind(vi));
const runAllTimers = withAct(vi.runAllTimers.bind(vi));
const runOnlyPendingTimers = withAct(vi.runOnlyPendingTimers.bind(vi));

const createFakeTimersHelpers = () => {
  return { advanceTimersByTime, runAllTimers, runOnlyPendingTimers };
};

type FakeTimersHelpers = ReturnType<typeof createFakeTimersHelpers>;
type RunFakeTimersCallback = (timers: FakeTimersHelpers) => void | Promise<void>;

export const runFakeTimers = <T extends RunFakeTimersCallback, R extends ReturnType<T>>(
  cb: T,
): R extends Promise<any> ? Promise<void> : void => {
  vi.useFakeTimers();
  const res = cb(createFakeTimersHelpers());
  if (res && 'then' in res) {
    // @ts-expect-error
    return res.finally(() => vi.useRealTimers());
  }
  vi.useRealTimers();
  // @ts-ignore
  return;
};
