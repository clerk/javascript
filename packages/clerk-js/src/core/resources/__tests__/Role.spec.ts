import { describe, expect, it, vi } from 'vitest';

import { Role } from '../Role';

const FIXED_DATE = new Date('2025-01-01T00:00:00.000Z');

describe('Role', () => {
  it('has the same initial properties', () => {
    vi.useFakeTimers();
    vi.setSystemTime(FIXED_DATE);

    const role = new Role({
      object: 'role',
      id: 'role_123',
      key: 'admin',
      name: 'Administrator',
      description: 'Full administrative access',
      permissions: [
        {
          object: 'permission',
          id: 'perm_1',
          key: 'users:read',
          name: 'Read Users',
          description: 'Can read user data',
          type: 'system',
          created_at: 1735689500000,
          updated_at: 1735689650000,
        },
        {
          object: 'permission',
          id: 'perm_2',
          key: 'users:write',
          name: 'Write Users',
          description: 'Can modify user data',
          type: 'system',
          created_at: 1735689500000,
          updated_at: 1735689650000,
        },
      ],
      created_at: 1735689500000,
      updated_at: 1735689650000,
    });

    expect(role).toMatchObject({
      id: 'role_123',
      key: 'admin',
      name: 'Administrator',
      description: 'Full administrative access',
      permissions: expect.arrayContaining([expect.any(Object), expect.any(Object)]),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
    });

    vi.useRealTimers();
  });
});

describe('Role Snapshots', () => {
  it('should match snapshot for role structure', () => {
    vi.useFakeTimers();
    vi.setSystemTime(FIXED_DATE);

    const role = new Role({
      object: 'role',
      id: 'role_123',
      key: 'admin',
      name: 'Administrator',
      description: 'Full administrative access',
      permissions: [
        {
          object: 'permission',
          id: 'perm_1',
          key: 'users:read',
          name: 'Read Users',
          description: 'Can read user data',
          type: 'system',
          created_at: 1735689500000,
          updated_at: 1735689650000,
        },
      ],
      created_at: 1735689500000,
      updated_at: 1735689650000,
    });

    expect(role).toMatchSnapshot();
    vi.useRealTimers();
  });

  it('should match snapshot for __internal_toSnapshot method', () => {
    const role = new Role({
      object: 'role',
      id: 'role_456',
      key: 'user',
      name: 'User',
      description: 'Basic user access',
      permissions: [],
      created_at: 1735689400000,
      updated_at: 1735689500000,
    });

    expect(role).toMatchSnapshot();
  });
});
