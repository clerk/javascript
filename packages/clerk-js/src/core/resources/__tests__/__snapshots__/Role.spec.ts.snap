// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Role Snapshots > should match snapshot for __internal_toSnapshot method 1`] = `
Role {
  "createdAt": 2024-12-31T23:56:40.000Z,
  "description": "Basic user access",
  "id": "role_456",
  "key": "user",
  "name": "User",
  "pathRoot": "",
  "permissions": [],
  "updatedAt": 2024-12-31T23:58:20.000Z,
}
`;

exports[`Role Snapshots > should match snapshot for role structure 1`] = `
Role {
  "createdAt": 2024-12-31T23:58:20.000Z,
  "description": "Full administrative access",
  "id": "role_123",
  "key": "admin",
  "name": "Administrator",
  "pathRoot": "",
  "permissions": [
    Permission {
      "createdAt": 2024-12-31T23:58:20.000Z,
      "description": "Can read user data",
      "id": "perm_1",
      "key": "users:read",
      "name": "Read Users",
      "pathRoot": "",
      "type": "system",
      "updatedAt": 2025-01-01T00:00:50.000Z,
    },
  ],
  "updatedAt": 2025-01-01T00:00:50.000Z,
}
`;
